src/snac_model.cpp

```diff
@@ -0,0 +1,220 @@
+    if (name == "alpha_out") {
+        snake_alpha = ggml_dup_tensor(ctx, tensor);
+        set_tensor(snake_alpha, tensor);
+    } else if (name == "in.weight") {
+        in_conv_kernel = ggml_dup_tensor(ctx, tensor);
+        set_tensor(in_conv_kernel, tensor);
+    } else if (name == "in.bias") {
+        in_conv_bias = ggml_dup_tensor(ctx, ggml_transpose(ctx, tensor));
+        set_tensor(in_conv_bias, tensor);
+    } else if (name == "up.weight") {
+        up_conv_kernel = ggml_dup_tensor(ctx, tensor);
+        set_tensor(up_conv_kernel, tensor);
+    } else if (name == "up.bias") {
+        up_conv_bias = ggml_dup_tensor(ctx, ggml_transpose(ctx, tensor));
+        set_tensor(up_conv_bias, tensor);
+    } else if (name == "final.weight") {
+        out_conv_kernel = ggml_dup_tensor(ctx, tensor);
+        set_tensor(out_conv_kernel, tensor);
+    } else if (name == "final.bias") {
+        out_conv_bias = ggml_dup_tensor(ctx, ggml_transpose(ctx, tensor));
+        set_tensor(out_conv_bias, tensor);
+    } else if (has_prefix(name, "layers")) {
+        auto pair = parse_layer_count(name);
+        int l = pair.first;
+        std::string lt_name = pair.second;
+        general_neural_audio_codec::assign_to_layer((tts_model *) this, layers[l], lt_name, tensor);
+    } else if (has_prefix(name, "quantizers")) {
+        auto pair = parse_layer_count(name);
+        int l = pair.first;
+        std::string lt_name = pair.second;
+        general_neural_audio_codec::assign_to_quantize_layer((tts_model *) this, quantizer_layers[l], lt_name, tensor);
+    }
+}
+
+static struct ggml_tensor * snac_build_audio_inputs(struct ggml_context * ctx, struct snac_context * sctx, size_t sequence_length, std::vector<general_neural_audio_codec::residual_vector_quantize_layer> layers) {
+    struct ggml_tensor * embd;
+    sctx->inp_tokens = ggml_new_tensor_1d(ctx, GGML_TYPE_I32, sequence_length / 4 + sequence_length / 2 + sequence_length);
+    ggml_set_input(sctx->inp_tokens);
+    size_t last_stride = 0;
+    for(int i = 0; i < sctx->model->n_heads; i++) {
+        auto quantize_layer = sctx->model->quantizer_layers[i];
+        struct ggml_tensor * inp_head = ggml_cont(ctx, ggml_view_1d(ctx, sctx->inp_tokens, sequence_length / sctx->model->repeats[i], last_stride));
+        last_stride += (sequence_length / sctx->model->repeats[i]) * ggml_element_size(sctx->inp_tokens);
+        struct ggml_tensor * code = general_neural_audio_codec::build_quantize_layer(ctx, inp_head, quantize_layer);
+        if (sctx->model->repeats[i] > 1) {
+            // this manipulation is equivalent to repeat_interleave against the first dimension of the tensor
+            code = ggml_repeat(ctx, ggml_cont_3d(ctx, code, 1, code->ne[0], code->ne[1]), ggml_new_tensor_3d(ctx, GGML_TYPE_F32, sctx->model->repeats[i], code->ne[0], sctx->model->embd)); //repeat_interleave(ctx, gf, code, ggml_cont(ctx, result), 0, sctx->model->repeats[i]);
```

Comment:
remove the commented section. cruft.

Reply:
Side note: repeat interleave is quite common and there are many different ways to achieve the operation in GGML. I should do a perf comparison and standardize an approach.